### DISCUSSION TITLE
Debugger

### DISCUSSION THREAD

**Discussion by jaredebird** [thingworx-field-work/ThingworxVSCodeProject#57] [Q&A] [answered]
I am now struggling to get the debugger working in VS Code.  I have the debug build uploaded that looks like this:

`
const __d = BMDebuggerRuntime.localDebugger(); 
const __dLogger = __d.getLogger();
__d.retainForService({name: "testDisplayAndon", filename: "D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts", scopeStack: []}); 
try { 
    var result = (function (logger) {const METHOD_NAME = "testDisplayAndon";

    var tempStartTime = (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-2'), new Date(1666984355738));
    var tempEndTime = (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-3'), new Date(1666984413941));
    var tempResponseStartTime = (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-4'), new Date(1666984396405));
    var inputAndon = (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-25'), { "LineDelaySeconds": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-5'), 0), "Description": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-6'), "Parts Running Out"), "EndTime": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-7'), tempEndTime), "HasParts": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-8'), true), "SubtypeUID": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-9'), 1), "StartTime": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-10'), tempStartTime), "TypeUID": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-11'), 1), "SiteUID": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-12'), 2), "Step": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-13'), 10), "SecondsOpen": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-14'), 58), "JobOrderUID": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-15'), 3387), "LineDelay": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-16'), false), "Station": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-17'), 3), "Line": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-18'), "3"), "UID": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-19'), 69853), "SerialNumber": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-20'), "75313"), "WorkDefUID": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-21'), 107715), "CycleContinue": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-22'), true), "Model": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-23'), "75313"), "ResponseStartTime": (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-24'), tempResponseStartTime) });
    var testAndon = (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-28'), Resources["InfoTableFunctions"].CreateInfoTableFromDataShape({
        infoTableName: (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-26'), "InfoTable"),
        dataShapeName: (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-27'), "AndonDS")
    }));
    __d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-30'), testAndon.AddRow(inputAndon);
    var displayAndon = (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-32'), Things.AndonHelper.DisplayAndon({ Andons: (__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-31'), testAndon) }));
    __d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-36'), logger.debug((__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-35'), JSON.stringify((__d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-34'), displayAndon.ToJSON()))));
    return __d.checkpoint('D:/ThingworxVSCodeProject-master/src/TestJared/JaredTestThing.ts-38'), true;
}).apply(me, [__dLogger]);
} 
finally { 
    __d.release(); 
}`

And I am connected in VS Code.  I can run the testDisplayAndon() function in the debug console and I get the result of true back, but it doesn't stop at the breakpoints.  Any suggestions?

**Comment by jaredebird** [comment]
Here is my package.json file
`{
	"start": "cross-env NODE_EXTRA_CA_CERTS=\"D:\\toro.pem\" node index.js",
	"name": "bm-thingworx-vscode",
	"version": "2.5.104",
	"description": "Develop in ThingWorx with a proper IDE.",
	"author": "Thingworx RoIcenter",
	"thingworxProjectName": "MyProject",
	"thingworxServer": "https://dev-thingworx.toro.com",
	"thingworxUser": "Administrator",
	"thingworxPassword": "Administrator12345",
	"thingworxAppKey": "**Removed**",
	"minimumThingWorxVersion": "6.0.0",
	"homepage": "https://github.com/BogdanMihaiciuc/",
	"autoUpdate": {
		"gitHubURL": "https://api.github.com/repos/ptc-iot-sharing/ThingworxVSCodeProject/releases/latest"
	},
	"repository": {
		"type": "git",
		"url": "https://github.com/ptc-iot-sharing/ThingworxVSCodeProject.git"
	},
	"scripts": {
		"test": "echo \"Error: no test specified\" && exit 1",
		"build": "twc build",
		"buildDebug": "twc build --debug",
		"upload": "twc upload",
		"uploadDebug": "twc upload --debug",
		"watch:declarations": "twc watch"
	},
	"license": "MIT",
	"files": [
		"lib",
		"build",
		"LICENSE"
	],
	"devDependencies": {
		"@types/node": "^8.10.11",
		"bm-thing-cli": "^1.4.5",
		"bm-thing-transformer": "^1.4.7"
	},
	"dependencies": {
		"gulp": "^4.0.2",
		"ssl-root-cas": "^1.3.1"
	}
}`

Here is my launch.json file.

`
{
    "version": "0.2.0",
    "configurations": [
		{
			"type": "thingworx-vscode",
			"request": "attach",
            "name": "Attach to Thingworx Test",
            "thingworxDomain": "dev-thingworx.toro.com",
            "thingworxPort": 8443,
            "thingworxAppKey": "***Removed***",
            "useSSL": true
		}
    ]
}
`

**Comment by BogdanMihaiciuc** [comment]
Ensure you did these:
 1. After installing the debug server extension, restart thingworx
 2. Ensure that the debug server subsystem state is running, if not start it
 3. After starting the debug server subsystem and whenever restarting thingworx, redeploy your project built in debug mode
 4. Make sure vscode doesn't report any errors when attaching; it will still show as attached after reporting errors
 5. Make sure the breakpoints are getting picked up by vscode. Inactive breakpoints appear gray, while active breakpoints are red and indicate exactly where at that line it will stop.

**Reply by BogdanMihaiciuc** [reply]
Also if you're on windows, make sure you are building with the latest cli tools/transformer. Previous versions had a bug where the debugger only worked on Linux and Mac.

**Reply by jaredebird** [reply]
I have done all that (multiple times) and still the problem continues.  Do I need to be working on a VS Code installation that is actually running on the same computer as Thingworx is installed on?  We have a dev server running Thingworx and I'm running VS Code on my local computer.

**Reply by jaredebird** [reply]
Also, if I run BMDebugServer.connectDebugger(), and then click attach to Thingworx in VS Code, the services seem to pause at the breakpoint, until I disconnect VS Code, but I don't get any debugging information in VS Code.  Just running DMDebugServer.connectDebugger() doesn't stop anything from running until I actually click the green play icon in VS Code.

**Reply by BogdanMihaiciuc** [reply]
With vscode attached, can you create a service with this code `BMDebuggerRuntime._connectedDebuggers`, or use the debug console to execute that? It should return a nonzero value. If it returns 0, it indicates there is a problem attaching properly.

It shouldn't matter where the server is located. Matching files is done based on the paths generated while building the project.

**Reply by BogdanMihaiciuc** [reply]
Also which version of Thingworx are you running and which version of the debug server extension?

**Reply by jaredebird** [reply]
We are running ThingWorx 9.3.4-b1190 and the bm-thingworx-debug-server extension is version 1.1.1.   

I do get a result of 0 when I run BMDebuggerRuntime._connectedDebuggers from the debug console.

**Reply by BogdanMihaiciuc** [reply]
Do you have any errors in your script or application logs related to the debugger?

**Reply by jaredebird** [reply]
I don't have any errors in the script log, but I do have the following error in the Application log.  
<img width="633" alt="App Log" src="https://user-images.githubusercontent.com/16650190/199266440-b8a2b8be-8c3b-4e17-acbd-eb3973de7a2b.png">

I do get a value of 1 for BMDebuggerRuntime._connectedDebuggers in the debug console after I manually run connectDebugger in Thingworx.

**Reply by BogdanMihaiciuc** [reply]
There shouldn't be a need to manually run that. The vscode debug extension runs it when attaching. I guess that the debug server somehow ended up in an invalid state. You could try to detach the vscode debugger then just restarting ThingWorx, re-uploading the debug project and attaching again. Do not invoke the debugger services manually and let me know if you still get 0 when checking the connected debuggers through the debug console.

**Comment by jaredebird** [comment, marked as answer]
OK, I did that and still the same thing.  I even removed the bm-thingworx-debug-server extension, restarted, reinstalled the server, restarted again, and tried to connect again. 
 
In launch.json, if I change `"thingworxDomain": "dev-thingworx.toro.com",` to `"thingworxDomain": "dev-thingworx.toro.com/",` (just added the forward slash after .com), I get the unable to verify first certificate error that I was seeing before.  Not sure if anything could be related there or not.

**Reply by BogdanMihaiciuc** [reply]
I'll try to test locally with a self signed cert to see if I can reproduce this behavior.

**Reply by jaredebird** [reply]
Thank you!

**Reply by jaredebird** [reply]
The problem was with my certificate.  I was able to get it to be accepted through VS Code by adding it in certmgr and it is working now.  Thanks again for your help!

**Reply by BogdanMihaiciuc** [reply]
Glad you got it working. I will need to update the documentation to specify that this is required for self signed certs in addition to just getting node to trust it.
